L72-core-google_default_credentials-extension.md

```diff
@@ -0,0 +1,62 @@
+Allow Call Credentials to be Specified in `grpc_google_default_credentials_create`
+----
+* Author(s): rbellevi
+* Approver: markdroth
+* Status: Draft
+* Implemented in: Core
+* Last updated: July 6th, 2020
+* Discussion at: (TODO)
+
+## Abstract
+
+This document proposes that `grpc_google_default_credentials_create` be
+amended to allow the user to specify their desired call credential, while
+also adding a `void* reserved` argument to lessen the burden of any such
+changes in the future.
+
+
+## Background
+
+The so-called "Google default credentials" created by the
+`grpc_google_default_credentials_create` function in Core enables connection to
+GCP via a combination of ALTS and SSL credentials, along with a special oauth2
+token that must assert the same identity as the channel-level ALTS credential.
+
+This oauth2 token attached by `grpc_google_default_credentials_create` is based
+on the default VM service account, but the auth libraries for wrapped languages
```

Comment:
Does this mean that the wrapped languages have a mechanism for changing the service account used in the channel-level ALTS credential?  How does that work?

Reply:
So I'm told. I don't have the context myself. @apolcyn should be able to provide it.

Reply:
Wrapped languages don't have a mechanism for changing the service account used in the ALTS credentials, but they don't really need to (as long as they're OK with authenticating as the default service account).

ALTS credentials by default authenticate the VM's default service account.  So long as the call credentials also authenticate the VM's default service account, then we're ok. And note that any call credentials which ultimately fetches the token from `http://metadata.google.internal/computeMetadata/v1/project/service-accounts/default/token` will authenticate the VM's default service account, creating a guarantee that we don't have a call-creds/ALTS identity mismatch.

Whether or not the wrapped language layers and above do the right thing, i.e. being sure that per-RPC tokens authenticate the VM's default service account, when potentially using ALTS, will be up to them though.